### PR TITLE
expose getBackendSrv() to plugins

### PR DESCRIPTION
**Pull request by ryantxu**
Simple PR to let plugins use:
```
import { BackendSrv, getBackendSrv } from 'grafana/app/core/services/backend_srv';
```
rather than needing to inject via angular

### CODE REVIEW

**Comment by dprokop** [comment]
Thank you for your contribution @ryantxu 

We are currently exploring different ways how we would like core services to be available for plugins. Our goal is to make grafana's building blocks accessible via npm packages. This work has just started and we are focusing on UI elements at the moment.

**Comment by ryantxu** [comment]
@dprokop any new thoughts on this?

I see two paths:
1. Everything gets cleanly migrated to @grafana/ui and external plugins can use them with good contracts
2. While we move stuff to @grafana/ui, some stuff gets added to the `exposeToPlugin` that lets people develop external plugins without getting too hacky.  However unless it is in @grafana/ui, there is no clear contract on it.

If `getBackendSrv()` is something you think will be moved soon, I can see waiting.  Otherwise I vote we put it in the middle ground and make it possible to work without an angular context.

**Comment by ryantxu** [comment]
@torkelo -- any thoughts on this?  Is there an easy near term path that lets things use getBackendSrv() without angular?

**Comment by torkelo** [comment]
* Wish there was a way expose this via @grafana/ui 
* So long term we will need a fetch style service/function exposed from @grafana/ui or a proper service interface with different methods for different types of request (datasource vs normal api requests). But could just combined into one fetch metod as well that just has a property to make a call into a datasourceRequest call. 

Question is we would probably want to rewite that so that it did not use angular $http and used axios or similar, so what to do in the meantime. 

We could expose this but would like to add some deprecation notice that it's not supported. I would hate to have new react plugins depend on this and not be able to remove it after we remove angular (in 2+ years)

**Comment by ryantxu** [comment]
> so what to do in the meantime.

exactly.  And I don't think this is the only place where it will be an issue.  My suggestion would be:

**Supported:**
Plugins built using:
```
"dependencies": {
  "@grafana/ui": "..."
},
```

**Unsupported/ Deprecated / on a path towards deprecation:**
Using a [local checkout of grafana](https://github.com/grafana/kentik-app/blob/master/build/prebuild.sh)

and/or depending on:
```
"devDependencies": {
  "@types/grafana": "github:CorpGlory/types-grafana"
  "grafana-sdk-mocks": "github:grafana/grafana-sdk-mocks",
},
```

This is already true for many things in `exposeToPlugin` for example:

```
exposeToPlugin('app/plugins/sdk', sdk);
exposeToPlugin('app/core/utils/datemath', datemath);
exposeToPlugin('app/core/utils/file_export', fileExport);
exposeToPlugin('app/core/utils/flatten', flatten);
exposeToPlugin('app/core/utils/kbn', kbn);
exposeToPlugin('app/core/utils/ticks', ticks);

exposeToPlugin('app/core/config', config);
exposeToPlugin('app/core/time_series', TimeSeries);
exposeToPlugin('app/core/time_series2', TimeSeries);
exposeToPlugin('app/core/table_model', TableModel);
exposeToPlugin('app/core/app_events', appEvents);
exposeToPlugin('app/core/core_module', coreModule);
...
```

----

The alternative immediate solution would be to make a `BackendSrv` interface in grafana/ui that the existing BackendSrv implements.  But I think that is just kicking the can, since we would likely not want to commit to `request(options: any)` as the main interface!

I'm looking for a way that we can make progress on more complex external react plugins without needing to wait for a full migration.  It is currently only possible to do simple visualisation plugins

**Comment by torkelo** [comment]
> But I think that is just kicking the can, since we would likely not want to commit to request(options: any) as the main interface!

No? was thinking that was what we wanted. (similar to fetch api). What is it you want from backendSrv beyond access to $http request? (with error / login handling). 

An alternative would be a fully typed api client with typed functions & responses for api resources. But that feels like something else. 

Just a bit hesitant to add this and have new react plugins depend on all the internal stuff in backendSrv. If we had simple interface / wrapper with only request & datasourceRequest then would feel better. I know angular plugins have full access today, so what's the harm.

**Comment by ryantxu** [comment]
closing in favor of #16533

**Comment by ryantxu** [comment]
reopened as an easy alternative to #16533

File: public/app/features/plugins/plugin_loader.ts [hunk 1]
```diff
@@ -93,6 +94,11 @@ exposeToPlugin('app/features/dashboard/impression_store', {
   __esModule: true,
 });
 
+exposeToPlugin('app/core/services/backend_srv', {
```
Comment (torkelo) [inline public/app/features/plugins/plugin_loader.ts]: Just add a comment here that this is added temporarily and will likely be replaced with something officially supported in the future. But is exposed to allow plugin authors a away to make http requests without having to wait for an official (via npm package / interface) solution. Only use the get/delete/post/patch/request methods, no other methods.

**Review by torkelo — APPROVED** [review]
Just one comment to add then you can merge this

File: public/app/features/plugins/plugin_loader.ts (hunk 1 above)
Comment (ryantxu) [reply public/app/features/plugins/plugin_loader.ts]: 👍 sounds good

**Pull request merged**